library: add modules (#57)
ceph/cephadm-ansible: doc/source/index.rst

.. contents::

Introduction
------------

cephadm-ansible is a collection of Ansible modules and playbooks to simplify
workflows that are not covered by [cephadm]_. The following workflows are covered
by three playbooks:

* cephadm-preflight.yml: Initial setup of hosts before bootstrapping the cluster
* cephadm-clients.yml: Setting up client hosts
* cephadm-purge-cluster.yml: Remove a Ceph cluster

In addition to that, several ansible modules are provided in order to let people writing their own playbooks.

.. Provided modules
.. ----------------

.. .. toctree::

> ..    :maxdepth: 1

Author: ```suggestion
```

.. contents::

Introduction
------------

cephadm-ansible is a collection of Ansible modules and playbooks to simplify
workflows that are not covered by [cephadm]_. The following workflows are covered
by three playbooks:

* cephadm-preflight.yml: Initial setup of hosts before bootstrapping the cluster
* cephadm-clients.yml: Setting up client hosts
* cephadm-purge-cluster.yml: Remove a Ceph cluster

> In addition to that, several ansible modules are provided in order to let people writing their own playbooks.

Reviewer: ```suggestion
Additionally, several ansible modules are provided in order to help users write their own playbooks.
```

.. [cephadm] https://docs.ceph.com/en/latest/cephadm/

Terminology
-----------

**admin host**

>   A host where the admin keyring and ceph config file is present. Although the admin host and the bootstrap host are usually the same host, it is possible to have multiple admin hosts later.\

Reviewer: ```suggestion
  A host where the admin keyring and ceph config file is present. Although the admin host and the bootstrap host are usually the same host, it is possible to have multiple admin hosts later. \
```
Not sure if it's just on my end but the spaces weren't rendering here when I tested locally
![image](https://user-images.githubusercontent.com/47704447/169865767-73dc8860-76bb-408a-a524-7f89ca2ddf93.png)

  ``cephadm`` will make a host become 'admin' when the label ``_admin`` is added to that host. (ie: ``ceph orch host label add <host> _admin``).\
  This hosts should be present in the group ``[admin]`` in the ansible inventory.\
  If for some reason you decide a host shouldn't be a 'admin host' anymore, you have to :

  * remove it from the group ``[admin]`` in the ansible inventory,
  * remove the admin keyring,
  * remove the ceph config file,
  * remove the ``_admin`` label. (ie ``ceph orch host label rm <host> _admin``)

**ansible host**
  The host where any cephadm-ansible playbook is run.

**bootstrap host**

>   The host where the ceph cluster will start. Unless you pass ``--skip-admin-label`` option to ``ceph bootstram`` command, this host will get the admin keyring and the ceph config file present, therefore, it should be considered as an 'admin host'.

Reviewer: ```suggestion
  The host where the ceph cluster will start. Unless you pass the ``--skip-admin-label`` option to the ``ceph bootstrap`` command, this host will receive the admin keyring and ceph config file and should be considered an 'admin host'.
```

>   This hosts should be present in the group ``[admin]`` in the ansible inventory.

Reviewer: ```suggestion
  This host should be present in the group ``[admin]`` in the ansible inventory.
```

Ansible inventory
-----------------

> The ansible inventory is a file where all the hosts intended to be part of the ceph cluster will be listed.\

Reviewer: seeing same as my other comment with the space not being added after the \

The most common format are INI or YAML.

> Although you probably want to keep it as simple as possible, you can organize your inventory and create groups, `cephadm-ansible` won't make any difference except for the following requirements:

Reviewer: ```suggestion
Although you probably want to keep it as simple as possible, you can organize your inventory and create groups, `cephadm-ansible` won't differentiate except for the following requirements:
```

* Client hosts must be defined in a dedicated group ``[clients]``.

> * Both ``cephadm-purge-cluster.yml`` and ``cephadm-clients.yml`` playbooks requires a group ``[admin]`` with at least one admin host (usually it will be the bootstrap node).

Reviewer: ```suggestion
* Both the ``cephadm-purge-cluster.yml`` and ``cephadm-clients.yml`` playbooks require a group ``[admin]`` with at least one admin host (usually it will be the bootstrap node).
```

.. note:: the name of the client group can be changed. In that case you have to set the variable `client_group`.

> Otherwise, you can create groups such as ``[monitors]``, ``[osds]``, ``[rgws]``, that might help you keep clarity in your inventory file and probably ease the ``--limit`` usage if you plan to use it to target a group of node only.

Reviewer: ```suggestion
Otherwise, you can create groups such as ``[monitors]``, ``[osds]``, ``[rgws]``, that might help you keep clarity in your inventory file and ease the ``--limit`` usage if you plan to use it to target a specific group of nodes only.
```

A basic inventory would look like following::

   # cat hosts
   ceph-mon1
   ceph-mon2
   ceph-mon3
   ceph-osd1
   ceph-osd2
   ceph-osd3
   ceph-mds1
   ceph-mds2
   ceph-rgw1
   ceph-rgw2

   [clients]
   ceph-client1
   ceph-client2
   ceph-client3

   [admin]
   ceph-mon1

Playbooks
---------

cephadm-preflight
=================

This playbook configures the Ceph repository.

> It also installs some prerequisites (podman, lvm2, chronyd, cephadm, ...)

Reviewer: do we want to note the issues with certain distros with this? Specifically thinking of the issues we saw with Fedora a while ago.

Author: yes, I think we should add a specific section for that

Usage::

   ansible-playbook -i <inventory host file> cephadm-preflight.yml

You can limit the execution to a set of hosts by using ``--limit`` option::

   ansible-playbook -i <inventory host file> cephadm-preflight.yml --limit <my_osd_group|my_node_name>

You can override variables using ``--extra-vars`` parameter::

   ansible-playbook -i <inventory host file> cephadm-preflight.yml --extra-vars "ceph_origin=rhcs"

If you plan to deploy client nodes, you must define a group called "clients" in your inventory::

   $ cat hosts
   node1
   node2
   node3

   [clients]
   client1
   client2
   client3
   node123

> Then you can run the playbook as shown above.

Reviewer: does this section about clients stuff need to be in this section? It seems like maybe it belongs in the section for the clients playbook.

Author: yes, it must be a leftover from when the clients setup was part of the preflight playbook

**default**
  "latest"

custom_repo_url
~~~~~~~~~~~~~~~
**description**
  The url of the repository when ``ceph_origin`` is 'custom'.

custom_repo_gpgkey
~~~~~~~~~~~~~~~~~~
**description**
  The url of the gpg key corresponding to the repository set in ``custom_repo_url`` when ``ceph_origin`` is 'custom'.

cephadm-purge
=============

This playbook purges a Ceph cluster managed with cephadm

You must define a group ``[admin]`` in your inventory with a node where
the admin keyring is present at ``/etc/ceph/ceph.client.admin.keyring``

Usage::

   ansible-playbook -i <inventory host file> cephadm-purge-cluster.yml -e fsid=<your fsid>

Options
+++++++

fsid
~~~~
**description**
  The fsid of the cluster.

cephadm-clients
===============

This playbooks distribute keyring and conf files to a set of client hosts.

Usage::

   ansible-playbook -i <inventory host file> cephadm-clients.yml -e fsid=<cluster fsid> -e keyring=<path to the keyring>

Options
+++++++

keyring
~~~~~~~~
**description**
  The full path name of the keyring file on the host (which should be admin[0]) which holds the key for the client to use

conf
~~~~
**description**

>   The full path name of the conf file on the (which should be admin[0]) host to use (undefined will trigger a minimal conf)

Reviewer: are there default values for the conf and keyring here?

Author: for `conf` : 

> (undefined will trigger a minimal conf)

`keyring` is a mandatory parameter, no default value

keyring_dest
~~~~~~~~~~~~
**description**
  The full path name of the destination where the keyring will be copied on the remote host. (default: /etc/ceph/ceph.keyring)

Modules
-------

Introduction
============

> cephadm-ansible provides several modules so it is possible to write playbooks in order to have some automation on top of cephadm/ceph orch.

Reviewer: ```suggestion
cephadm-ansible provides several modules to make it easier to write playbooks around cephadm/ceph orch.
```

The idea is to let you write your own playbooks so you can have a solution that really fits with your needs.

> At the moment only most important tasks are supported. It means that any operation uncovered would have to be done either with either ``command`` or ``shell`` Ansible tasks in your playbook.

Reviewer: ```suggestion
At the moment only the most important tasks are supported. This means that any operation not covered would have to be done either with either the ``command`` or ``shell`` Ansible tasks in your playbook.
```

  If set to 'absent', it will remove the host specified in 'name'.
  If set to 'drain', it will schedule to remove all daemons from the host specified in 'name'.

ceph_config
+++++++++++

``fsid``
  The fsid of the Ceph cluster to interact with.
``action``
  Whether to get or set the parameter specified in 'option'.
``who``
  Which daemon the configuration should be set to.
``option``
  Name of the parameter to be set.
``value``
  Value of the parameter to set.

ceph_orch_apply
+++++++++++++++

``fsid``
  The fsid of the Ceph cluster to interact with.
``spec``
  The service spec to apply.

ceph_orch_daemon
++++++++++++++++

``fsid``
  The fsid of the Ceph cluster to interact with.
``state``
  The desired state of the service specified in 'name'.
  If 'started', it ensures the service is started.
  If 'stopped', it ensures the service is stopped.
  If 'restarted', it will restart the service.
``service_id``
  The id of the service.
``service_type``
  The type of the service.

Samples
=======

This shows how the supported modules can be used in a playbook.

> This doesn't cover the pre-requisites steps (preflight, ...) so it implies all requirements satisfied (podman, lvm2,...).

Reviewer: ```suggestion
This doesn't cover the pre-requisites steps (preflight, ...) so it assumes all requirements are satisfied (podman, lvm2,...).
```

> It assumes your "bootstrap host" (or "admin host") can ssh other hosts with root user without password.

Reviewer: ```suggestion
It assumes your "bootstrap host" (or "admin host") can ssh to other hosts with root user without password.
```

Bootstrap and add some hosts::

   # cat hosts
   ceph-mon1 monitor_address=10.10.10.101 labels="['_admin', 'mon', 'mgr']"
   ceph-mon2 labels="['mon', 'mgr']"
   ceph-mon3 labels="['mon', 'mgr']"
   ceph-osd1 labels="['osd']"
   ceph-osd2 labels="['osd']"
   ceph-osd3 labels="['osd']"
   # cat site.yml
   ---
   - name: bootstrap the cluster
     hosts: ceph-node01
     become: true
     gather_facts: false
     tasks:
       - name: bootstrap initial cluster
         cephadm_bootstrap:
           mon_ip: "{{ monitor_address }}"

   - name: add more hosts
     hosts: all
     become: true
     gather_facts: true
     tasks:
       - name: add hosts to the cluster
         ceph_orch_host:
           name: "{{ ansible_facts['hostname'] }}"
           address: "{{ ansible_facts['default_ipv4']['address'] }}"
           labels: "{{ labels }}"
         delegate_to: ceph-mon1

   - name: deploy osd service
     hosts: ceph-node01
     become: true
     gather_facts: false
     tasks:
       - name: apply osd spec
         ceph_orch_apply:
         spec: |

>             service_type: osd
>             service_id: osd
>             placement:
>             label: osd
>             spec:
>             data_devices:
>                   all: true

Reviewer: ```suggestion
            service_type: osd
            service_id: osd
            placement:
                  label: osd
            spec:
                  data_devices:
                        all: true
```
The spec will fail to apply without the indentation I think

>    - name: change osd_memory_target option for osd.0

Reviewer: note that in some newer version we turn on osd memory autotuning by default during bootstrap so if somebody sets this config option it might just get overwritten. Not sure if that means a different example would be better here.

Author: > Not sure if that means a different example would be better here.

I think so
